File: graphql/udemy1-graphql17/GraphQL-HTTP-authentication/src/main/java/com/accounts/config/SecurityConfig.java
package com.accounts.config;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.InMemoryUserDetailsManager;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;

import static org.springframework.security.config.Customizer.withDefaults;


@Configuration
@EnableWebSecurity
@EnableMethodSecurity
public class SecurityConfig {

    @Autowired
    private AuthEntryPointBasicAuth unauthorizedHandler;

    @Value("${app.user.admin}")
    private String adminUser;

    @Value("${app.user.admin.password}")
    private String adminPassword;

    @Value("${app.user.user}")
    private String appUser;

    @Value("${app.user.user.password}")
    private String appUserPassword;

    @Bean
    SecurityFilterChain defaultSecurityFilterChain(HttpSecurity http) throws Exception {

        http.csrf(csrf -> csrf.disable());

        // Add a custom filter to check for the Authorization header
        http.addFilterBefore(new AuthorizationHeaderFilter(), BasicAuthenticationFilter.class);

        http.authorizeHttpRequests(authorizeRequests ->
                authorizeRequests
                        .requestMatchers("/graphql").permitAll()
                        .anyRequest().authenticated());

        http.exceptionHandling(exception -> exception.authenticationEntryPoint(unauthorizedHandler));

        http.httpBasic(withDefaults());

        return http.build();
    }

    @Bean
    public UserDetailsService userDetailsService() {
        UserDetails user1 = User.withUsername(appUser)
                .password(passwordEncoder().encode(appUserPassword))
                .roles("USER")
                .build();

        UserDetails user2 = User.withUsername(adminUser)
                .password(passwordEncoder().encode(adminPassword))
                .roles("ADMIN")
                .build();

        return new InMemoryUserDetailsManager(user1, user2);
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}

File: graphql/graphqlex/dgs-web/src/main/kotlin/com/yscorp/dgsweb/security/SecurityConfig.kt
package com.yscorp.dgsweb.security

import com.yscorp.dgsweb.domain.UserRepository
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configurers.AuthorizeHttpRequestsConfigurer.AuthorizationManagerRequestMatcherRegistry
import org.springframework.security.web.SecurityFilterChain


@Configuration
@EnableMethodSecurity(securedEnabled = true)
class SecurityConfig(
    private val userzRepository: UserRepository
) {

    @Bean
    @Throws(Exception::class)
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        val authProvider: ProblemzAuthenticationProvider = ProblemzAuthenticationProvider(userzRepository)
//
//        http.authenticationProvider(authProvider)
//        http.csrf().disable().authorizeHttpRequests(
//            Customizer { auth: AuthorizationManagerRequestMatcherRegistry ->
//                auth.anyRequest().authenticated()
//            }
//        )

        return http.build()
    }
}
